frontend/apps/reader/modules/readeruserhyph.lua

```diff
@@ -0,0 +1,357 @@
+local ConfirmBox = require("ui/widget/confirmbox")
+local DataStorage = require("datastorage")
+local Event = require("ui/event")
+local FFIUtil = require("ffi/util")
+local InfoMessage = require("ui/widget/infomessage")
+local InputDialog = require("ui/widget/inputdialog")
+local MultiConfirmBox = require("ui/widget/multiconfirmbox")
+local UIManager = require("ui/uimanager")
+local WidgetContainer = require("ui/widget/container/widgetcontainer")
+local TimeVal = require("ui/timeval")
+local lfs = require("libs/libkoreader-lfs")
+local logger = require("logger")
+local _ = require("gettext")
+local T = require("ffi/util").template
+
+local ReaderUserHyph = WidgetContainer:new{
+    -- messages from setUserHyphenationDict
+    USER_DICT_RELOAD = 0,
+    USER_DICT_NOCHANGE = 1,
+    USER_DICT_MALFORMED = 2,
+    USER_DICT_ERROR = 3,
+
+    -- work with malformed dictionary entries, some hyphenations might not work
+    malformed_accepted = false,
+}
+
+-----------------------------------------------
+-- Helper for the UI functions
+-----------------------------------------------
+
+-- returns path to the user dictionary
+function ReaderUserHyph:getDictionaryPath()
+    return FFIUtil.joinPath(DataStorage:getSettingsDir(),
+        "user-" .. self.ui.document:getTextMainLangDefaultHyphDictionary():gsub(".pattern$", "") .. ".hyph")
+end
+
+
+local load_error_text = _([[Error loading user dictionary:
+%1
+
+%2]])
+
+-- Load the user dictionary suitable for the actual language
+-- if reload==true, force a reload
+-- Unload is done automatically when a new dictionary is loaded.
+function ReaderUserHyph:loadDictionary(name, reload)
+    if G_reader_settings:isTrue("hyph_user_dict") and lfs.attributes(name, "mode") == "file" then
+        local load_message = self.ui.document:setUserHyphenationDict(name, reload)
+        -- this should only happen, if a user edits a dictionary by hand or the user messed
+        -- with the dictionary file by hand. -> Warning and disable.
+        if load_message == self.USER_DICT_ERROR then
+            UIManager:show(InfoMessage:new{
+                text = T(load_error_text, name, _("Dictionary is not sorted alphabetically.\n\nDictionary is disabled now.")),
+            })
+            G_reader_settings:saveSetting("hyph_user_dict", false)
+        elseif load_message == self.USER_DICT_MALFORMED and not self.malformed_accepted then
+            UIManager:show(ConfirmBox:new{
+                text =  T(load_error_text, name, _("At least one dictionary entry is malformed.\n\nDo you want to work with that dictionary.")),
```

Comment:
I'm not even sure what this means.

Reply:
This means, that the user has edited the user dict by hand (PC or texteditor) and has made a mistake (e.g. `danger;dn-ger`) like skipped a letter, one letter to much ... 
So if the user knows what he has done, he can correct it. (If not, we have a hint where to look for.)

Such entries are skipped by crengine, so you can work with the rest of the dictionary.

A suggestion for a different wording (in my Denglich) would be:
`text =  T(_("Error loading user dictioary:\n%1\nAt least one dictionary entry is invalid.\nYou can work with this incomplete dictionary, but some hyphenations might not be as you expect."),name),`

Reply:
If that's only in the case of manual editing, may be a `logger.warn(english string)` is enough? No need to make the UI complicated (and useless work for translators) for a non-nominal use case?

Reply:
I understand your point. Both errors can only show up, if the user has done something with the file by hand.

Technically: 
*) The first error is severe. If the dict is not sorted correct you get randomly (depending on the dict and the word) expected user hyphenations or the build in hyphenations (TexHyph, AlgoHyph, SoftHyph). You never get wrong ones from the dictionary.

*) In the second case the hyphenation for some (wrong edited) hyphenations are dropped.

We could fold the two messages/cases into one, telling the user he has borked something. Maybe `Dictionary %1 has been edited outside of KOReader. There are some errors (sorting, incorrect enterd hyphenations) in it. Dictionary is disabled now."
and throw some more detailed logger warnings.

Reply:
I'd say keep 2 InfoMessage, one for each case, with a generic message (don't assume it has been edited outside of KOReader, it may have from inside, or you may have some bugs in our frontend editing code :):
- "The user hyphenation dictionary %1 is corrupted (not sorted, ...), not using it"
- "The user hyphenation dictionary %1 has corrupted entries, using only valid ones" (no ConfirmBox, use it even if incomplete, the user gets the InfoMessage and can decide to go back to working on fixing it).
And you can logger.warn() more detailed info about what is wrong.